### DISCUSSION TITLE
[DONE] Why does `DIDUrl::join` takes ownership of `self` instead of borrowing it?

### DISCUSSION THREAD

**Discussion by arnauorriols** [iotaledger/identity.rs#805] [Technical Discussion]
I'm referring to [`DIDUrl::join`](https://github.com/iotaledger/identity.rs/blob/c51071a764ccc159b2119420f308466058e4c857/identity-did/src/did/did_url.rs#L396-L407). If I'm not mistaken, the method does not _need_ ownership of the instance, nor it is customary that "join" methods take it (as can be seen in [`Path::join`](https://doc.rust-lang.org/std/path/struct.Path.html#method.join) or [`Slice::join`](https://doc.rust-lang.org/std/primitive.slice.html#method.join)). This forces client code to redundantly clone the `DIDUrl`. Is it maybe a semantical reason?

**Comment by cycraig** [comment]
I could be wrong but I think the general reasoning was not to hide clones in functions, similar to why `DID::join` consumes self and requires a preceding `clone` or `to_url` call.

Personal opinion but given the behaviour of existing APIs you mentioned and the focus shift towards being more developer-friendly and easy-to-use, I would vote for changing `DID::join` and `DIDUrl::join` to borrow self.

**Reply by arnauorriols** [reply]
> I could be wrong but I think the general reasoning was not to hide clones in functions

Don't get me wrong, if borrowing self would force a clone within the method, then I would get behind taking self by value, as stablished by [the Rust guidelines](https://rust-lang.github.io/api-guidelines/flexibility.html#c-caller-control). But by the looks of it (and I might have overlooked something) `self` is never needed by value within `DIDUrl::join`. It's only borrowed in `self.to_string()`. 

> similar to why DID::join consumes self and requires a preceding clone or to_url call.

I would consider `DID::join` an special case, because it's just a shortcut to `did.into_url().join(value)`, and this indeed reasonably needs self by value to convert into a `DIDUrl`

It's worth noting that the underlying [`did_url::DID::join`](https://docs.rs/did_url/latest/did_url/struct.DID.html#method.join) method borrows `self` instead of owning it (if API consistency is a matter of concern)

**Comment by cycraig** [comment]
Changed in PR #871, thanks for the suggestion.
